brave-kafka/src/main/java/smartthings/brave/kafka/producers/TracingProducerInterceptor.java

```diff
@@ -0,0 +1,101 @@
+
+/**
+ * Copyright 2016-2017 SmartThings
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package smartthings.brave.kafka.producers;
+
+import brave.Span;
+import brave.Tracer;
+import brave.propagation.TraceContext;
+import com.google.protobuf.ByteString;
+import org.apache.kafka.clients.ClientUtils;
+import org.apache.kafka.clients.producer.ProducerInterceptor;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.config.ConfigException;
+import smartthings.brave.kafka.EnvelopeProtos;
+import smartthings.brave.kafka.consumers.*;
+import zipkin.Constants;
+import zipkin.Endpoint;
+
+import java.math.BigInteger;
+import java.net.InetSocketAddress;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class TracingProducerInterceptor<K> implements ProducerInterceptor<K, byte[]> {
+
+  private Tracer tracer;
+  private SpanNameProvider<K> nameProvider;
+  private final Endpoint kafkaEndpoint;
+
+
+  public TracingProducerInterceptor() {
+    kafkaEndpoint = Endpoint.builder().serviceName("Kafka").build();
+  }
+
+  @Override
+  public ProducerRecord<K, byte[]> onSend(ProducerRecord<K, byte[]> record) {
+
+    // TODO: how to join local span saved in ClientSpanThreadBinder?
+    Span span = tracer.newTrace()
```

Comment:
I think you know but this needs to be `newChild(parent.context())`.  Without that the span won't be joined to its parent in the UI.  Since we are dealing with a Kafka Interceptor and there are no gaurenties as to which thread this will be called on I'm not sure how we can ensure the correct parent context is used.

Maybe the Interceptor pattern isn't going to work in this case and we need to do something with a proxy of some sort or other byte code manipulation?